lenskit-core/src/main/java/org/lenskit/results/AbstractResult.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * LensKit, an open source recommender systems toolkit.
+ * Copyright 2010-2014 LensKit Contributors.  See CONTRIBUTORS.md.
+ * Work on LensKit has been funded by the National Science Foundation under
+ * grants IIS 05-34939, 08-08692, 08-12148, and 10-17697.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.results;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.lenskit.api.Result;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Base class for basic result types.  It provides storage for the ID and score, as well as helper methods for hashing
+ * and equality checking.  This type does not directly enforce immutability, but subclasses should be immutable.
+ */
+public abstract class AbstractResult implements Result {
+    protected long id;
+    protected double score;
+
+    /**
+     * Create a new result.
+     * @param id The result ID.
+     * @param score The result score.
+     */
+    protected AbstractResult(long id, double score) {
+        this.id = id;
+        this.score = score;
+    }
+
+    /**
+     * Create a new, uninitialized result.
+     */
+    protected AbstractResult() {}
+
+    @Override
+    public long getId() {
+        return id;
+    }
+
+    @Override
+    public double getScore() {
+        return score;
+    }
+
+    @Override
+    public boolean hasScore() {
+        return !Double.isNaN(score);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * The default implementation simply casts the result to type `type` if possible.
+     */
+    @Override
+    public <T extends Result> T as(@Nonnull Class<T> type) {
+        if (type.isInstance(this)) {
+            return type.cast(this);
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * The default implementation simply calls {@link #as(Class)}.
+     */
+    @Override
+    public <T extends Result> T find(@Nonnull Class<T> type) {
+        return as(type);
```

Comment:
This is what I expect most Lenskit users mental model of as/find to be (should they even have a mental model).